zerver/lib/test_runner.py

```diff
@@ -431,11 +432,13 @@ def run_tests(
         # a Django connection to be rolled back mid-test.
         with get_sqlalchemy_connection():
             result = self.run_suite(suite)
+            assert isinstance(result, TextTestResult)
         self.teardown_test_environment()
         failed = self.suite_result(suite, result)
         if not failed:
             write_instrumentation_reports(full_suite=full_suite, include_webhooks=include_webhooks)
-        return failed, result.failed_tests
+        self.last_failed_tests = result.failed_tests
```

Comment:
Why is this called `last_failed_tests`? It seems like this function is only called once per process, I think?

Reply:
Also, would it be clearer to just have the caller pass a mutable object in this function, and then have this mutate that, rather than adding new state to the `Runner` class?  Not sure, but this implementation kinda feels like an abstraction violation.

Reply:
I see. It is named this way to emphasize that this attribute is affected by a side-effect. Passing a mutable instance should work just fine.

Reply:
Updated.